test/unit-test/twcc_manager/twcc_manager_utest.c

```diff
@@ -28,30 +112,167 @@ void test_twccAddPacket( void )
+    for( i = 0; i < TWCC_PACKET_INFO_ARRAY_LENGTH; i++ )
+    {
+        result = RtcpTwccManager_FindPacketInfo( &( twccManager ),
+                                                 seqNum + ( uint16_t ) i,
+                                                 &( foundTwccPacketInfo ) );
+
+        TEST_ASSERT_EQUAL( RTCP_TWCC_MANAGER_RESULT_OK,
+                           result );
+        TEST_ASSERT_EQUAL( sentTime + ( uint64_t ) i,
+                           foundTwccPacketInfo.localSentTime );
+        TEST_ASSERT_EQUAL( packetSize + ( size_t ) i,
+                           foundTwccPacketInfo.packetSize );
+        TEST_ASSERT_EQUAL( seqNum + ( uint16_t ) i,
+                           foundTwccPacketInfo.packetSeqNum );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Validate Twcc Manager Add packet functionality for Overflow.
+ */
+void test_twccAddPacket_Overflow( void )
+{
+    uint16_t seqNum = 256;
+    uint32_t i;
+    RtcpTwccManager_t twccManager;
+    RtcpTwccManagerResult_t result;
+    TwccPacketInfo_t twccPacketInfo = { 0 };
+    TwccPacketInfo_t foundTwccPacketInfo;
+    size_t packetSize = ( rand() % ( 1000 ) );
+    uint64_t sentTime = time( NULL );
+
+    result = RtcpTwccManager_Init( &( twccManager ),
+                                   &( twccPacketInfoArray[ 0 ] ),
+                                   TWCC_PACKET_INFO_ARRAY_LENGTH );
+
+    TEST_ASSERT_EQUAL( RTCP_TWCC_MANAGER_RESULT_OK,
+                       result );
+
+    for( i = 0; i < TWCC_PACKET_INFO_ARRAY_LENGTH; i++ )
+    {
+        twccPacketInfo.packetSize = packetSize + ( size_t ) i;
+        twccPacketInfo.localSentTime = sentTime + ( uint64_t ) i;
+        twccPacketInfo.packetSeqNum = seqNum + ( uint16_t ) i;
+
+        result = RtcpTwccManager_AddPacketInfo( &( twccManager ),
+                                                &( twccPacketInfo ) );
+
+        TEST_ASSERT_EQUAL( RTCP_TWCC_MANAGER_RESULT_OK,
+                           result );
+        TEST_ASSERT_EQUAL( i + 1,
+                           twccManager.count );
+    }
+
+    twccPacketInfo.packetSize = packetSize + ( size_t ) i;
+    twccPacketInfo.localSentTime = sentTime + ( uint64_t ) i;
+    twccPacketInfo.packetSeqNum = seqNum + ( uint16_t ) i;
+
+    result = RtcpTwccManager_AddPacketInfo( &( twccManager ),
+                                            &( twccPacketInfo ) );
+
+    TEST_ASSERT_EQUAL( RTCP_TWCC_MANAGER_RESULT_OK,
+                       result );
+    TEST_ASSERT_EQUAL( TWCC_PACKET_INFO_ARRAY_LENGTH,
+                       twccManager.count );
+
+    for( i = 1; i <= TWCC_PACKET_INFO_ARRAY_LENGTH; i++ )
+    {
+        result = RtcpTwccManager_FindPacketInfo( &( twccManager ),
+                                                 seqNum + ( uint16_t ) i,
+                                                 &( foundTwccPacketInfo ) );
+
+        TEST_ASSERT_EQUAL( RTCP_TWCC_MANAGER_RESULT_OK,
```

Comment:
In this test case, the length starts from 0, but you check it from 1.
Does that mean 0 packet is overwritten by the overflowed packet?
Could you add a check for that packet as well in this test case?

Reply:
So in this case we are the packetInfoarray[] is of size `TWCC_PACKET_INFO_ARRAY_LENGTH` ( for simplicity say `n`) Now for the overflow condition we add n + 1 packets. Since the addpacket api auto deletes if the package age is more than TWCC estimated time window, in this case the packet age is not that , so it will keep on filling till n entires are filled. the moment `n+1 `entry is added , it deletes the first ( i.e oldest package ). You are getting confused why i is starting from 1, not 0. The packetInfoarray will start from index 0 only which will get called inside the` RtcpTwccManager_FindPacketInfo() `api, but to call this api we are starting the index from i = 1, since I am checking that count of array is n and all the packets from **1 -- n+ 1** can be retrieved. So definitely 0th packet got deleted.